sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/ContinuousDataSourceRDD.scala

```diff
@@ -0,0 +1,153 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.datasources.v2
+
+import java.util.concurrent.atomic.{AtomicBoolean, AtomicLong}
+
+import scala.collection.JavaConverters._
+
+import org.apache.spark._
+import org.apache.spark.rdd.RDD
+import org.apache.spark.rpc.RpcEndpointRef
+import org.apache.spark.sql.Row
+import org.apache.spark.sql.catalyst.expressions.UnsafeRow
+import org.apache.spark.sql.execution.streaming._
+import org.apache.spark.sql.execution.streaming.continuous._
+import org.apache.spark.sql.sources.v2.reader.{ContinuousDataReader, DataReader, ReadTask}
+import org.apache.spark.sql.streaming.ProcessingTime
+import org.apache.spark.util.SystemClock
+
+// This can't be a structural type because we need to mutate its atomic vars.
+class ContinuousDataSourceRDDIter(
+    epochEndpoint: RpcEndpointRef,
+    context: TaskContext,
+    reader: DataReader[UnsafeRow])
+  extends Iterator[UnsafeRow] {
+
+  private[this] var valuePrepared = false
+
+  var currentEpoch = {
+    val startEpoch = context.getLocalProperty(ContinuousExecution.START_EPOCH_KEY).toLong
+    val offsetJson = reader match {
+      case r: RowToUnsafeDataReader =>
+        r.rowReader.asInstanceOf[ContinuousDataReader[Row]].getOffset().json
+      case _ => throw new IllegalStateException("must have ContinuousDataReader[Row]")
+    }
+    epochEndpoint.send(ReportPartitionOffset(
+      context.partitionId(), startEpoch, offsetJson))
+
+    startEpoch
+  }
+
+  var outputMarker = new AtomicBoolean(false)
+
+  override def hasNext: Boolean = {
+    if (context.isInterrupted()) {
+      // On interrupt, end the iterator so the writer can stop itself, but don't report a new epoch.
+      return false
```

Comment:
I actually originally had InterruptibleIterator here. My idea was that we'd allow the current record to bubble up to the writer before ending the task, in order to achieve clean shutdown (since continuous processing tasks will never end naturally). Is there a reason the task must be killed immediately?